### PR TITLE
Initial Commit

### PR DESCRIPTION
**Pull request by ausgreen**
Austin Green
Mohamed Hamza

### CODE REVIEW

File: tic_tac_toe/game.py [hunk 1]
```diff
@@ -24,18 +28,31 @@ def _position_is_valid(position):
 
     Returns True if given position is valid, False otherwise.
     """
-    pass
-
+    if isinstance(position,int):
```
Comment (jsymons) [inline tic_tac_toe/game.py:31]: Why are you specifically checking if position is an integer? You're expecting a tuple, so you should check if it is *not* a tuple: `if not isinstance(position, tuple):` The only reason this is passing is because our test values happen to fit into the parameters of your if statements. These are the invalid entries we test against that you're code is catching for the wrong reasons:

- `"something"` is being caught because of your `len(position) > 2`, if this were instead `"Hi"` or some other two character string the test would fail
- `False` is being caught because technically a `boolean` is an instance of `int`. (Strictly speaking True is 1 and False is 0)

File: tic_tac_toe/game.py [hunk 2]
```diff
@@ -63,22 +83,50 @@ def _check_winning_combinations(board, player):
...
-    pass
+    first_row = ((0, 0), (0, 1), (0, 2))
+    second_row = ((1, 0), (1, 1), (1, 2))
+    third_row = ((2, 0), (2, 1), (2, 2))
+    first_col = ((0, 0), (1, 0), (2, 0))
+    second_col = ((0, 1), (1, 1), (2, 1))
+    third_col = ((0, 2), (1, 2), (2, 2))
+    diag_1 = ((0, 0), (1, 1), (2, 2))
+    diag_2 = ((0, 2), (1, 1), (2, 0))
+    winning_combinations  = [first_row, second_row, third_row, first_col, second_col, third_col, diag_1, diag_2]
```
Comment (jsymons) [inline tic_tac_toe/game.py:94]: This could be simplified as follows:
```
winning_combinations= [((0,0), (0,1), (0,2)), # rows
                       ((1,0), (1,1), (1,2)),
                       ((2,0), (2,1), (2,2)),
                       ((0,0), (1,0), (2,0)), # cols
                       ((0,1), (1,1), (2,1)),
                       ((0,2), (1,2), (2,2)),
                       ((0,0), (1,1), (2,2)), # diags
                       ((0,2), (1,1), (2,0))]

File: tic_tac_toe/game.py [hunk 3]
```diff
@@ -87,18 +135,54 @@ def move(game, player, position):
...
+    game['board'][position[0]][position[1]] = player
+    if player == game['player1']:
+        game['next_turn'] = game['player2']
+    elif player == game['player2']:
+        game['next_turn'] = game['player1']
+    
+    
+    # check game endings
+    game['winner'] = get_winner(game)
+    if game['winner'] == game['player1']:
```
Comment (jsymons) [inline tic_tac_toe/game.py:161]: This can be simplified to `if game['winner']:` and then just using `raise GameOver('"{}" wins!'.format(player))` You could also use `game['winner']` instead of `player`, however since you're always checking at the end of a player's turn if the game has been won the winner will always be the current player.

File: tic_tac_toe/game.py [hunk 4]
```diff
@@ -87,18 +135,54 @@ def move(game, player, position):
...
+        raise GameOver('"X" wins!')
+    elif game['winner'] == game['player2']:
+        raise GameOver('"O" wins!')
+    elif _board_is_full(game['board']) and game['winner'] == None:
```
Comment (jsymons) [inline tic_tac_toe/game.py:165]: Don't need to check if there is a game winner at this point as if there was your previous if statements would catch it.

**Review by jsymons — COMMENTED** [review]
It's important to do a `git status` before you do your `add` and `commit` to make sure you're only adding what you want to. You've inadvertently added a bunch of c9 temp files to your commit.

File: tic_tac_toe/game.py (hunk 1 above)
Comment (ausgreen) [reply tic_tac_toe/game.py:31]: Hey @jsymons thanks for taking the time to look through my code.  Yeah, I'm not really sure what I was thinking about the first error catch.  In a flurry, I was probably thinking about the next problem I had to solve.  

It turns out that the last two conditions do a pretty good job at catching strings.  

if position[0] > 2 or position[1] < 0:
        return False
    if position[1] > 2 or position[1] < 0:
        return False



>>> x = ("hi")
>>> isinstance(x, int)
False
>>> len(x)
2
>>> x[0] > 2
True
>>> x[0] < 0
False
>>> x[1] > 2
True
>>> x[1] < 0
False
>>> 

I guess I just got lucky.   Perhaps the tests should be written to be more picky?  I just wrote it quick and saw it passed tests then moved on.

File: tic_tac_toe/game.py (hunk 2 above)
Comment (ausgreen) [reply tic_tac_toe/game.py:94]: @jsymons That is how I had written it before, but found that it was easier for me to visualize/write the logic with the tuples named this way.

File: tic_tac_toe/game.py (hunk 3 above)
Comment (ausgreen) [reply tic_tac_toe/game.py:161]: @jsymons I knew there was a better way of doing this.

File: tic_tac_toe/game.py (hunk 1 above)
Comment (jsymons) [reply tic_tac_toe/game.py:31]: That won't actually work on python3.
```
   x = 'hi'
   x[0] > 1
Traceback (most recent call last):
  File "python", line 1, in <module>
TypeError: unorderable types: str() > int()
```
As for the tests there's a limit to how many edge cases we should be testing for. You could conceivably write an if statement for every single bit of test data we could toss in, but it would still be missing the point. Using `isinstance(position, tuple)` would catch anything we could toss that isn't a tuple(or a class that extends tuple) and doesn't rely on getting lucky with the data.
